File: backend/app/schemas/replay.py
from datetime import datetime
from typing import Any

from pydantic import BaseModel


class ReplayFrame(BaseModel):
    step_id: int
    step_name: str
    agent_role: str | None = None
    status: str
    started_at: datetime | None = None
    completed_at: datetime | None = None
    duration_ms: int | None = None
    input_payload: dict[str, Any] | None = None
    output_payload: dict[str, Any] | None = None
    error_message: str | None = None
    model_calls_count: int = 0
    tokens_used: int = 0
    cost_usd: float = 0.0


class TaskReplayResponse(BaseModel):
    task_id: int
    title: str
    status: str
    total_duration_ms: int | None = None
    total_cost_usd: float = 0.0
    frames: list[ReplayFrame]


class ResumeStepRequest(BaseModel):
    step_id: int
    custom_instruction: str | None = None

File: frontend/src/types/replay.ts
export interface ReplayFrame {
  step_id: number
  step_name: string
  agent_role: string | null
  status: string
  started_at: string | null
  completed_at: string | null
  duration_ms: number | null
  input_payload: Record<string, unknown> | null
  output_payload: Record<string, unknown> | null
  error_message: string | null
  model_calls_count: number
  tokens_used: number
  cost_usd: number
}

export interface TaskReplayResponse {
  task_id: number
  title: string
  status: string
  total_duration_ms: number | null
  total_cost_usd: number
  frames: ReplayFrame[]
}
